### ISSUE TITLE
tmbprofile/tmbroot issues with REML fits

### ISSUE THREAD

**Issue by bbolker** [glmmTMB/glmmTMB#501]
I'm having trouble running `TMB::tmbprofile` or `TMB::tmbroot` with glmmTMB fits with REML=TRUE. (I'm guessing this is a glmmTMB rather than a TMB problem because `tmbprofile` and `tmbroot` work fine with plain old random effects ...

It makes sense that we should only be able to run profile/uniroot on the first parameter (since there is really only one "fixed" parameter in REML models), but that doesn't seem to work either.  I'm fine with saying "just don't do that" if there's a theoretical reason it's not sound, but I wouldn't want to stop people computing profile CIs on REML fits if the problem is on our end somewhere.

```
library(glmmTMB)
fm0 <- glmmTMB(Murder~Illiteracy+Population+Area+`HS Grad`,
               data=as.data.frame(state.x77), REML = TRUE)

tmbprofile(fm0$obj, name=1) ## all NA
try(tmbprofile(fm0$obj, name=2))
> Error in intI(i, n = di[margin], dn = dn[[margin]], give.dn = FALSE) : 
>   'NA' indices are not (yet?) supported for sparse Matrices

TMB::tmbroot(fm0$obj, name=1) ## NA/NA
try(TMB::tmbroot(fm0$obj, name=2)) ## same error as above

data(sleepstudy, package="lme4")
fm1 <- glmmTMB(Reaction~Days+(Days|Subject), sleepstudy)
t1 <- tmbprofile(fm1$obj, name=1, trace=FALSE) ## fine
t2 <- tmbprofile(fm1$obj, name=2, trace=FALSE) ## fine
t1R <- TMB::tmbroot(fm1$obj, name=1, trace=FALSE) ## fine
t2R <- TMB::tmbroot(fm1$obj, name=2, trace=FALSE) ## fine
```

**Comment by bwiernik** [comment]
Has there been any progress on this issue? Is profiling REML variance components generally a poor idea (e.g., is the profiling implementation in lmer not a good idea)? I'd love to be able to switch entirely to glmmTMB instead of lmer, but the loss of profile CIs with REML is a barrier.

**Comment by bbolker** [comment]
I wouldn't say it's a poor idea at all. (Among other things, Doug Bates would probably the first person *not* to implement something if it weren't statistically sound.)  But I have to admit I don't fully know what's going on here. @kaskr ... ?

**Comment by kaskr** [comment]
I don't see a problem - this works as expected:

```
tmbprofile(fm0$obj, name=1)
```

Perhaps it failed previously due to:
https://github.com/kaskr/adcomp/commit/b96566332fc36893b433db6f8e5c3d6dacee17bb
?

Of course one cannot do

```
tmbprofile(fm0$obj, name=2)
```

because there's only one variance parameter in the model:

```
> fm0$obj$par
betad 
    0
```

**Comment by bbolker** [comment]
Yes, I think I was confused. The only issue here that could be improved (perhaps) is to make the error message more informative! (Should double-check to see how `profile()` works in glmmTMB for REML-fitted models ... users shouldn't be expected to have to figure out the numerical position in the (random-excluding) parameter vector corresponding to a term of interest ...)
